### ISSUE TITLE
No rows returned by useVirtualizer in unit tests

### ISSUE THREAD

**Issue by densk1** [TanStack/virtual#641]
### Describe the bug

In browser table rows are rendered as expected. In unit tests `useVirtualizer` returns an empty array from `rowVirtualizer.getVirtualItems()`. The bug is introduce with changes introduced by `v3.0.0-beta.63`. It also exists in the latest release `v3.0.1`

I think it was introduced by #598 [here](https://github.com/TanStack/virtual/pull/598/files#diff-85d1b85f3d4e501779ca41e51cd38cf7f61a2b7ffddaf958cd7e79c2787c36f1R592-R599)

### Your minimal, reproducible example

below

### Steps to reproduce

This unit test will pass for `v3.0.0-beta.62` and fail for any version there after.

```ts
import { useVirtualizer } from "@tanstack/react-virtual";
import { render, screen } from "@testing-library/react";
import { styled } from "@mui/material";
import { useCallback, useRef } from "react";

const items = [
  { id: 1, name: "foo" },
  { id: 2, name: "bar" },
];

const Parent = styled("div")({ height: "100%", width: "100%", overflow: "auto" });

const Inner = styled("div")({ width: "100%", position: "relative" });

const Row = styled("div")({ position: "absolute", top: 0, left: 0, width: "100%" });

function VirtualTableComponent() {
  const parentRef = useRef<HTMLDivElement>(null);

  const rowVirtualizer = useVirtualizer({
    count: items.length,
    getScrollElement: () => parentRef.current,
    estimateSize: useCallback(() => 56, []),
  });

  return (
    <Parent ref={parentRef}>
      <Inner sx={{ height: `${rowVirtualizer.getTotalSize()}px` }}>
        {rowVirtualizer.getVirtualItems().map((virtualRow) => (
          <Row
            key={virtualRow.index}
            data-index={virtualRow.index}
            ref={rowVirtualizer.measureElement}
            style={{ transform: `translateY(${virtualRow.start}px)` }}
          >
            <div>{items[virtualRow.index].name}</div>
          </Row>
        ))}
      </Inner>
    </Parent>
  );
}

describe("@tanstack/react-virtual", () => {
  it("renders rows", () => {
    render(<VirtualTableComponent />);

    expect(screen.getByText(items[0].name)).toBeInTheDocument();
    expect(screen.getByText(items[1].name)).toBeInTheDocument();
  });
});
```

### Expected behavior

As a user I expect the provided unit test to pass but it fails.

### How often does this bug happen?

None

### Platform

macOS

### tanstack-virtual version

v3.0.1

### TypeScript version

v4.9.5

### Terms & Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I understand that if my bug cannot be reliable reproduced in a debuggable environment, it will probably not be fixed and this issue may even be closed.

**Comment by piecyk** [comment]
@densk1 that is correct behaviour, it's because in unit test elements don't have height or width. One option is to mock the getBoundingClientRect for the scroller element to return some height.

**Comment by densk1** [comment]
Ah ok, thanks! So it won't matter if set height on parent CSS because getBoundingClientRect won't return it anyway?

Is there a challenge with mocking getBoundingClientRect that it needs to be mocked for both the parent, inner and rows themselves?

**Comment by ak274** [comment]
> @densk1 that is correct behaviour, it's because in unit test elements don't have height or width. One option is to mock the getBoundingClientRect for the scroller element to return some height.

I'm facing the same issue `rowVirtualizer.getVirtualItems()` is always empty. Is there any example how to write unit tests for this?

**Comment by luchiago** [comment]
Same issue here using `3.0.1`. The proposed solution `One option is to mock the getBoundingClientRect for the scroller element to return some height.` worked on my test suite.

```javascript
Element.prototype.getBoundingClientRect = jest.fn(() => {
        return {
            width: 120,
            height: 120,
            top: 0,
            left: 0,
            bottom: 0,
            right: 0,
        }
    });
```

**Comment by zroux93** [comment]
@luchiago When I tried this, it changed the sizes of all HTML elements (both the container and the option elements), so my test list only rendered 2 options (presumably a visible one and the option below it) even though I passed more options to it. Is this how your tests are written or do you have a workaround for this?

**Comment by luchiago** [comment]
In my case, the whole test file was failing for the same reason, the same as you. The tests passed when I put a `beforeEach` with this workaround. But before it started failing, I only expected an array with two virtual rows. That may be why it worked for me.

**Comment by saomartinho** [comment]
@luchiago, can you put a example? Because a follow your suggestion, and put `beforeEach`, but still return 0 virtual rows.

**Comment by luchiago** [comment]
@saomartinho can you share the code problem here?

**Comment by istvandesign** [comment]
I think the issue here is that we should have unit tests in the examples, it's not a mature enough library without examples on how to do testing.

**Comment by luchiago** [comment]
@istvandesign it's a similar approach to what I did, but you did it with more options. I agree it's missing a documentation on how to mock, but I believe it should be an issue to investigate.

**Comment by istvandesign** [comment]
> @istvandesign it's a similar approach to what I did, but you did it with more options. I agree it's missing a documentation on how to mock, but I believe it should be an issue to investigate.

That's because I am using TypeScript.

**Comment by piecyk** [comment]
> Documentation should contain at least how to mock getBoundingClientRect in jest with JS and TypeScript to test with TanStack/virtual.

Yes i agree, we should include it. Nevertheless mocking Element.prototype is enough, for more complex scenarios we can even return different values, for example like here if we add `aria-label` on scroller div of virtualizer 

```tsx
const getDOMRect = (width: number, height: number) => ({
  width,
  height,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
  x: 0,
  y: 0,
  toJSON: () => {},
})

beforeEach(() => {
  Element.prototype.getBoundingClientRect = jest.fn(function () {
    if (this.getAttribute('aria-label') === 'Virtual list') {
      return getDOMRect(500, 500)
    }
    return getDOMRect(0, 0)
  })
})
```

> Do we have some other way to do this, like mocking some function inside the library ?

Other option would be to create context provider that would pass your own observeElementRect for test, something like

```tsx
const VirtualizerObserveElementRect = React.createContext<
  ((instance: Virtualizer<any, any>, cb: (rect: { width: number; height: number }) => void) => void) | undefined
>(undefined)

// in test wrap with provider 

const Mock = ({ children }: { children: React.ReactNode }) => {
  return (
    <VirtualizerObserveElementRect.Provider
      value={(_, cb) => {
        cb({ height: 200, width: 200 })
      }}
    >
      {children}
    </VirtualizerObserveElementRect.Provider>
  )
}

// and in component 

const observeElementRect = React.useContext(VirtualizerObserveElementRect)
const virtualizer = useVirtualizer({
  count: 1000,
  observeElementRect,
  estimateSize: () => 25,
  getScrollElement: () => null,
})
```

**Comment by ipsips** [comment]
i am facing the same problem but not in simulated test environment:

it seems to me that if getScrollElement returns null, then getVirtualItems returns an empty array.
and it seems that getScrollElement is called during initial render and if you've chosen to return a ref object's current value (which is expected to point to DOM element) then you are not doing React correctly because:

- "During the first render, the DOM nodes have not yet been created, so ref.current will be null." [from the docs](https://react.dev/learn/manipulating-the-dom-with-refs#when-react-attaches-the-refs)
- "Do not write or read ref.current during rendering." [also from the docs](https://react.dev/reference/react/useRef#referencing-a-value-with-a-ref)

coming up with a workaround to this should be trivial, however using a ref object to refer to a DOM node does seem like to most natural approach in getScrollElement.

**Comment by denchen** [comment]
Sorry to dredge up an old topic, but I had been using the mocking of `getBoundingClientRect()` from https://github.com/TanStack/virtual/issues/641#issuecomment-1902239048  successfully, but this no longer works with @TanStack/virtual@3.13.7 (published on May 5th). I believe it's due to this change -- TanStack/virtual#986 -- with no longer using `getBoundingClientRect()`.

Width this change, I'm unsure with how to properly mock the calculated width / height.

**Comment by shniubobo** [comment]
> Sorry to dredge up an old topic, but I had been using the mocking of `getBoundingClientRect()` from [#641 (comment)](https://github.com/TanStack/virtual/issues/641#issuecomment-1902239048) successfully, but this no longer works with @TanStack/virtual@3.13.7 (published on May 5th). I believe it's due to this change -- [#986](https://github.com/TanStack/virtual/pull/986) -- with no longer using `getBoundingClientRect()`.
> 
> Width this change, I'm unsure with how to properly mock the calculated width / height.
> 
> **EDIT:** The following is a new workaround that passes my tests:
> 
>   beforeEach(() => {
>     Object.defineProperty(HTMLElement.prototype, 'offsetHeight', {
>       value: 800
>     });
>     Object.defineProperty(HTMLElement.prototype, 'offsetWidth', {
>       value: 800
>     });
>   });

Thank you very much for the new workaround! In my case, it has worked even without `beforeEach`.

**Comment by piecyk** [comment]
Yes, that's correct! We’ve transitioned to using offsetHeight and offsetWidth for measuring the scroller's size. To mock these values properly, you’ll need to adjust your mocks accordingly. This definitely should be mentioned in the documentation to avoid confusion. I'll make sure to bring it up.

**Comment by nataliethistime** [comment]
For anyone interested, another way of solving this is to pass in a custom element rect observer that returns the value you need. We had to take this approach as we were mocking `ResizeObserver` to do nothing due to other parts of our application.

The usage of this function is [documented here](https://tanstack.com/virtual/latest/docs/api/virtualizer#observeelementrect) and you can also refer to the [original implementation](https://github.com/TanStack/virtual/blob/94761ff2720ea4f20d0b77453d78aa4aabd1ba9f/packages/virtual-core/src/index.ts#L67) if it helps.

```typescript
jest.mock<typeof import('@tanstack/react-virtual')>('@tanstack/react-virtual', () => {
  const actual = jest.requireActual<typeof import('@tanstack/react-virtual')>('@tanstack/react-virtual')
  return {
    ...actual,
    useVirtualizer(options) {
      return actual.useVirtualizer({
        ...options,
        observeElementRect: (instance, callback) => {
          if (instance.scrollElement) callback({ width: 10_000, height: 10_000 })
        },
      })
    },
  }
})
```

**Comment by hashcode-adrien** [comment]
I wrote a custom VirtualListBox for joyui `Autocomplete` which uses `@tanstack/react-virtual` instead of `react-window`.

I share the code to make it work as expected with **vitest** :

```tsx
import { Autocomplete } from '@mui/joy';
import { render, screen } from '@testing-library/react';
import { expect, test } from 'vitest';
import VirtualListBox from '../joyui/VirtualListBox';

// Save original measurement APIs so we can restore them after the test.
const originalGetBoundingClientRect = Element.prototype.getBoundingClientRect;

test('renders Autocomplete with VirtualListBox and check if virtualbox_option2 is visible', async () => {
  try {
    Object.defineProperties(HTMLElement.prototype, {
      clientHeight: {
        configurable: true,
        get() {
          return 500;
        }
      },
      clientWidth: {
        configurable: true,
        get() {
          return 500;
        }
      },
      offsetHeight: {
        configurable: true,
        get() {
          return 500;
        }
      },
      offsetWidth: {
        configurable: true,
        get() {
          return 500;
        }
      }
    });
    render(
      <Autocomplete
        sx={{ height: 500 }}
        options={["virtualbox_option1", "virtualbox_option2"]}
        slots={{ listbox: VirtualListBox }}
        open={true}
        renderOption={(props, option: string) => [props, option] as React.ReactNode}
      />
    );

    // wait for an option to appear; findByText will retry until it finds the element
    const opt = await screen.findByText(/virtualbox_option2/);
    expect(opt).toBeTruthy();
  } finally {
    // restore the original to avoid side effects for other tests
    Element.prototype.getBoundingClientRect = originalGetBoundingClientRect;
  }
});

```
